Fix CTAS for non-hdfs storages, also fixes multi storage cases

vikrambohra opened the pull request:
## Summary
This PR introduces the following changes 
1. Fix CTAS for non-hdfs storage types 
While extracting UUID from a snapshot, the code constructs the database path excluding the endpoint (scheme) when checking if it is a prefix of the manifestList that is part of the snapshot

Example
ManifestList (from snapshot): s3://bucket-name/database/table-uuid/file.avro
Database prefix: bucket-name/database (not a prefix of above)

Fix: Strip the endpoint(scheme) from the manifest list by resolving the correct storage from the tableLocation

After fix
ManifestList stripped : bucket-name/database/table-uuid/file.avro
Database prefix: bucket-name/database (is a prefix of above)

2. Fix multiple storage scenario
There are assumptions in code that storage is always cluster default. This fails when the default is a storage without scheme (hdfs) and the db.table is being stored in a storage with scheme (S3, BlobFs)

Fix: Resolve the correct storage for by extracting the tableLocation from table props and checking the scheme (endpoint)

3. Adds a method to storage interface to check if the tableLocation exists 
4. Add a method to StorageClient to check if a specified path exists.

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [x] Bug Fixes
- [ ] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [x] Refactoring
- [ ] Documentation
- [x] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done
<!--- Check any relevant boxes with "x" -->

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

1. Updated TableUUIDGeneratorTest 
2. Added TableUUIDGeneratorMultiStorageTest
3. Ran Docker setup for both s3 and hdfs

```
HDFS (CTAS from empty and non empty table)

scala> spark.sql("CREATE TABLE openhouse.db.t1 (name string)")
res11: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t1").show(100, false)
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                              |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|openhouse.tableUri                        |LocalHadoopCluster.db.t1                                                                                           |
|openhouse.tableUUID                       |1ed4be6b-c7e3-489f-87b1-1f9d698302be                                                                               |
|current-snapshot-id                       |none                                                                                                               |
|openhouse.lastModifiedTime                |1733279081615                                                                                                      |
|write.metadata.delete-after-commit.enabled|true                                                                                                               |
|write.metadata.previous-versions-max      |28                                                                                                                 |
|openhouse.tableLocation                   |/data/openhouse/db/t1-1ed4be6b-c7e3-489f-87b1-1f9d698302be/00000-e5e8e340-44ad-452a-80b0-a128ea43cc5b.metadata.json|
|openhouse.tableCreator                    |openhouse                                                                                                          |
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                      |
|policies                                  |                                                                                                                   |
|openhouse.tableId                         |t1                                                                                                                 |
|openhouse.databaseId                      |db                                                                                                                 |
|openhouse.creationTime                    |1733279081615                                                                                                      |
|openhouse.clusterId                       |LocalHadoopCluster                                                                                                 |
|format                                    |iceberg/orc                                                                                                        |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                    |
|write.format.default                      |orc                                                                                                                |
|write.parquet.compression-codec           |zstd                                                                                                               |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas1 AS SELECT * FROM openhouse.db.t1")
res13: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SELECT * FROM openhouse.db.ctas1").show()
+----+
|name|
+----+
+----+


scala> spark.sql("DROP TABLE openhouse.db.ctas1").show()
++
||
++
++


scala> spark.sql("CREATE TABLE openhouse.db.t1 (name string)")
res0: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t1").show(100, false)
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                              |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|openhouse.tableUri                        |LocalHadoopCluster.db.t1                                                                                           |
|current-snapshot-id                       |none                                                                                                               |
|openhouse.creationTime                    |1733278950420                                                                                                      |
|write.metadata.delete-after-commit.enabled|true                                                                                                               |
|openhouse.lastModifiedTime                |1733278950420                                                                                                      |
|write.metadata.previous-versions-max      |28                                                                                                                 |
|openhouse.tableCreator                    |openhouse                                                                                                          |
|openhouse.tableLocation                   |/data/openhouse/db/t1-6d9e0feb-4fd0-4b32-aafb-c1062965119e/00000-f4ef6f3e-3ccb-47a4-9c0a-be0c7434b0bb.metadata.json|
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                      |
|policies                                  |                                                                                                                   |
|openhouse.tableId                         |t1                                                                                                                 |
|openhouse.databaseId                      |db                                                                                                                 |
|openhouse.clusterId                       |LocalHadoopCluster                                                                                                 |
|format                                    |iceberg/orc                                                                                                        |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                    |
|write.format.default                      |orc                                                                                                                |
|write.parquet.compression-codec           |zstd                                                                                                               |
|openhouse.tableUUID                       |6d9e0feb-4fd0-4b32-aafb-c1062965119e                                                                               |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("INSERT INTO openhouse.db.t1 values ('value1')")
res2: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.t1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas1 AS SELECT * FROM openhouse.db.t1")
res4: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.ctas1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("DROP TABLE openhouse.db.ctas1").show()
++
||
++
++


S3 (CTAS from empty and non empty table)

scala> spark.sql("CREATE TABLE openhouse.db.t1 (name string)")
res6: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t1").show(100, false)
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                                    |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|openhouse.tableUUID                       |b0e20fcd-f23b-45c4-a0c1-b5c95dfc4fb1                                                                                     |
|openhouse.clusterId                       |LocalS3Cluster                                                                                                           |
|current-snapshot-id                       |none                                                                                                                     |
|openhouse.tableLocation                   |s3://openhouse-bucket/db/t1-b0e20fcd-f23b-45c4-a0c1-b5c95dfc4fb1/00000-17a150a2-8297-4509-ba38-4932f5ce7601.metadata.json|
|write.metadata.delete-after-commit.enabled|true                                                                                                                     |
|write.metadata.previous-versions-max      |28                                                                                                                       |
|openhouse.tableCreator                    |openhouse                                                                                                                |
|openhouse.tableUri                        |LocalS3Cluster.db.t1                                                                                                     |
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                            |
|policies                                  |                                                                                                                         |
|openhouse.tableId                         |t1                                                                                                                       |
|openhouse.lastModifiedTime                |1733278559045                                                                                                            |
|openhouse.creationTime                    |1733278559045                                                                                                            |
|openhouse.databaseId                      |db                                                                                                                       |
|format                                    |iceberg/orc                                                                                                              |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                          |
|write.format.default                      |orc                                                                                                                      |
|write.parquet.compression-codec           |zstd                                                                                                                     |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas1 AS SELECT * FROM openhouse.db.t1")
res8: org.apache.spark.sql.DataFrame = []

scala> spark.sql("DROP TABLE openhouse.db.ctas1").show()
++
||
++
++

scala> spark.sql("CREATE TABLE openhouse.db.t2 (name string)")
res10: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t2").show(100, false)
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                                    |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|openhouse.clusterId                       |LocalS3Cluster                                                                                                           |
|current-snapshot-id                       |none                                                                                                                     |
|openhouse.tableUUID                       |26182fc1-1896-4f91-9c48-bee6f210f843                                                                                     |
|openhouse.tableUri                        |LocalS3Cluster.db.t2                                                                                                     |
|openhouse.tableId                         |t2                                                                                                                       |
|write.metadata.delete-after-commit.enabled|true                                                                                                                     |
|write.metadata.previous-versions-max      |28                                                                                                                       |
|openhouse.tableCreator                    |openhouse                                                                                                                |
|openhouse.lastModifiedTime                |1733278628329                                                                                                            |
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                            |
|policies                                  |                                                                                                                         |
|openhouse.tableLocation                   |s3://openhouse-bucket/db/t2-26182fc1-1896-4f91-9c48-bee6f210f843/00000-37c7b221-d247-4791-a90e-3d0ca2bbee43.metadata.json|
|openhouse.databaseId                      |db                                                                                                                       |
|format                                    |iceberg/orc                                                                                                              |
|openhouse.creationTime                    |1733278628329                                                                                                            |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                          |
|write.format.default                      |orc                                                                                                                      |
|write.parquet.compression-codec           |zstd                                                                                                                     |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("INSERT INTO openhouse.db.t2 values ('value1')")
res13: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SELECT * FROM openhouse.db.t2").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas2 AS SELECT * FROM openhouse.db.t2")
res15: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SELECT * FROM openhouse.db.ctas2").show()
+------+
|  name|
+------+
|value1|
+------+
```
Testing Updated (Dec 5th 2024)

```
// S3

scala> spark.sql("CREATE TABLE openhouse.db.t1 (name string)")
res1: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t1").show(100, false)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                                    |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|openhouse.creationTime                    |1733417270322                                                                                                            |
|openhouse.clusterId                       |LocalS3Cluster                                                                                                           |
|current-snapshot-id                       |none                                                                                                                     |
|openhouse.tableUUID                       |43c6ce15-da7a-44c6-8d45-82adb2fbc2d6                                                                                     |
|openhouse.lastModifiedTime                |1733417270322                                                                                                            |
|write.metadata.delete-after-commit.enabled|true                                                                                                                     |
|openhouse.tableLocation                   |s3://openhouse-bucket/db/t1-43c6ce15-da7a-44c6-8d45-82adb2fbc2d6/00000-06ac5f04-1ffe-485e-9b05-00d3dd78adcd.metadata.json|
|write.metadata.previous-versions-max      |28                                                                                                                       |
|openhouse.tableCreator                    |openhouse                                                                                                                |
|openhouse.tableUri                        |LocalS3Cluster.db.t1                                                                                                     |
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                            |
|policies                                  |                                                                                                                         |
|openhouse.tableId                         |t1                                                                                                                       |
|openhouse.databaseId                      |db                                                                                                                       |
|format                                    |iceberg/orc                                                                                                              |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                          |
|write.format.default                      |orc                                                                                                                      |
|write.parquet.compression-codec           |zstd                                                                                                                     |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("INSERT INTO openhouse.db.t1 values ('value1')")
res3: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.t1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas1 AS SELECT * FROM openhouse.db.t1")
res5: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SELECT * FROM openhouse.db.ctas1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("DROP TABLE openhouse.db.ctas1").show()
++
||
++
++

// hdfs

scala> spark.sql("CREATE TABLE openhouse.db.t1 (name string)")
res0: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.t1").show(100, false)
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|key                                       |value                                                                                                              |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
|openhouse.tableUri                        |LocalHadoopCluster.db.t1                                                                                           |
|openhouse.tableLocation                   |/data/openhouse/db/t1-d78b7c49-44bf-4b55-ae95-c88bf27505b2/00000-7d8e2d06-ea77-4552-9131-51821fb267f5.metadata.json|
|current-snapshot-id                       |none                                                                                                               |
|openhouse.creationTime                    |1733417449464                                                                                                      |
|openhouse.tableUUID                       |d78b7c49-44bf-4b55-ae95-c88bf27505b2                                                                               |
|write.metadata.delete-after-commit.enabled|true                                                                                                               |
|write.metadata.previous-versions-max      |28                                                                                                                 |
|openhouse.tableCreator                    |openhouse                                                                                                          |
|openhouse.tableType                       |PRIMARY_TABLE                                                                                                      |
|policies                                  |                                                                                                                   |
|openhouse.tableId                         |t1                                                                                                                 |
|openhouse.lastModifiedTime                |1733417449464                                                                                                      |
|openhouse.databaseId                      |db                                                                                                                 |
|openhouse.clusterId                       |LocalHadoopCluster                                                                                                 |
|format                                    |iceberg/orc                                                                                                        |
|openhouse.tableVersion                    |INITIAL_VERSION                                                                                                    |
|write.format.default                      |orc                                                                                                                |
|write.parquet.compression-codec           |zstd                                                                                                               |
+------------------------------------------+-------------------------------------------------------------------------------------------------------------------+


scala> spark.sql("INSERT INTO openhouse.db.t1 values ('value1')")
res2: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.t1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("CREATE TABLE openhouse.db.ctas1 AS SELECT * FROM openhouse.db.t1")
res4: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.ctas1").show()
+------+
|  name|
+------+
|value1|
+------+


scala> spark.sql("DROP TABLE openhouse.db.ctas1").show()
++
||
++
++
```
For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

HotSushi commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -38,6 +36,8 @@ public class TableUUIDGenerator {
 
   @Autowired StorageManager storageManager;
 
+  @Autowired OpenHouseInternalCatalog catalog;
```
Lets avoid adding `OpenHouseInternalCatalog` at this layer and resolve this purely using `Storage` apis. Reasoning: We may onboard non-iceberg table specs, and internalcatalog introduces only-iceberg semantics.

jainlavina replied:
+1

vikrambohra replied:
done

HotSushi commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -3,24 +3,22 @@
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.Path;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.iceberg.catalog.TableIdentifier;
```
nit: same as comment below, lets avoid introducing iceberg at this layer.

jainlavina replied:
+1

vikrambohra replied:
donr

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/hdfs/HdfsStorageClient.java:
```diff
@@ -48,4 +49,14 @@ public FileSystem getNativeClient() {
+  public boolean pathExists(String path) {
+    try {
+      return fs.exists(new Path(path));
+    } catch (IOException e) {
```
does fs.exists throw an IOException for non-existing paths? thats weird

jainlavina replied:
Looks like it returns true or false: https://github.com/apache/hadoop/blob/cd2cffe73f909a106ba47653acf525220f2665cf/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L1860

But could throw an IOException if getFileStatus() throws an IOException.

I am wondering if it could throw an IOException in scenarios like HDFS is unavailable but the file exists?

HotSushi commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -143,11 +143,13 @@ private void validatePathOfProvidedRequest(
-            storageManager, dbIdFromProps, tblIdFromProps, tableUUIDProperty);
-    if (TableType.REPLICA_TABLE != tableType && !doesPathExist(previousPath)) {
-      log.error("Previous tableLocation: {} doesn't exist", previousPath);
+    Storage storage = catalog.resolveStorage(TableIdentifier.of(dbIdFromProps, tblIdFromProps));
```
resolveStorage will use `SelectStorage` if the entry doesn't exist in HTS. That code is not guaranteed to be idempotent during migration, lets avoid using selectStorage again in the consecutive COMMIT/PutSnapshot call.

Can we instead use tableLocation and derive storage from there? ie:
if s3://a/b/c -> S3Storage
if abs://a/b/c -> ABSStorage
if a/b/c -> HDFSStorage

vikrambohra replied:
done. Please check. there is a catch though, because both local and hdfs storage have same paths

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java:
```diff
@@ -76,4 +76,20 @@ public String allocateTableLocation(
+   * @return true if table location exists else false
+   */
+  @Override
+  public boolean tableLocationExists(
```
❤️

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageClient.java:
```diff
@@ -47,4 +47,12 @@ public interface StorageClient<T> {
+   * @param path
+   * @return true if path exists else false
+   */
+  boolean pathExists(String path);
```
❤️

HotSushi reviewed (changes requested):
Agree with Storage API changes. One feedback on not introducing iceberg at rest layer.

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/Storage.java:
```diff
@@ -70,4 +70,16 @@ public interface Storage {
+   * @param tableId the table id of the table
+   * @param tableUUID the UUID of the table
+   * @param tableCreator the creator of the table
+   * @return true if table location exists else false
```
Ditto

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageClient.java:
```diff
@@ -47,4 +47,12 @@ public interface StorageClient<T> {
+   * @param path
+   * @return true if path exists else false
+   */
+  boolean pathExists(String path);
```
Can you please add comment about the expected format of path? Should it include scheme or not? Is it URI-style path? Or is it absolute path starting from root folder or bucket?

vikrambohra replied:
done

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/adls/AdlsStorageClient.java:
```diff
@@ -64,4 +64,10 @@ public DataLakeFileClient getNativeClient() {
+
+  @Override
+  public boolean pathExists(String path) {
+    // TODO: Support pathExists on ADLS
```
Use exists()?
https://learn.microsoft.com/en-us/java/api/com.azure.storage.file.datalake.datalakepathclient?view=azure-java-stable#com-azure-storage-file-datalake-datalakepathclient-exists()

vikrambohra replied:
Can we do this as part of separate PR? DLFC is not initialized: https://github.com/linkedin/openhouse/issues/148

jainlavina replied:
Sure

jainlavina replied:
nit: should we then throw not implement exception instead?

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:
```diff
@@ -55,4 +56,11 @@ public S3Client getNativeClient() {
   }
+
+  @Override
+  public boolean pathExists(String path) {
```
What is the intended usage of this existence check?
1. Do you want to check whether the bucket exists or not?
2. Or, do you want to check whether any blob for the table has been created yet or not?

The two checks above are very different. For 1 you should check existence of bucket. For 2 you should check existence of objects with a prefix that way you are doing it now.

I would imagine that you intend to check that the bucket or root location has been created and hence 1 and not 2. Is that right?

vikrambohra replied:
The intention is to whether directory exists so it is check 2.
Current code passes the absolute path until table 
-> blobfs://bucket_name/data/openhouse/db/t-uuid (blobfs)
-> s3://bucket_name/data/openhouse/db/t-uuid (s3)
-> /data/openhouse/db/t-uuid (hdfs)

vikrambohra replied:
This is no longer valid. See https://github.com/linkedin/openhouse/pull/256#issuecomment-2492717671

vikrambohra commented:
@HotSushi @jainlavina Addressed the comments. Some changes in the latest commit
1. Removed tableLocationExists() from the storage api - It is longer required for us to construct the table location since we fetch it from table properties. 
2. Changed the pathExists in storageClient api to fileExists to be explicit about the check since we are now checking the absolute path of metadata,json file (table location) instead of only the path till table directory.

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java:
```diff
@@ -76,4 +76,20 @@ public String allocateTableLocation(
+   * @param tableId the table id of the table
+   * @param tableUUID the UUID of the table
+   * @param tableCreator the creator of the table
+   * @return true if table location exists else false
```
done

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/hdfs/HdfsStorageClient.java:
```diff
@@ -48,4 +49,20 @@ public FileSystem getNativeClient() {
+   * @return true if path exists else false
+   */
+  @Override
+  public boolean fileExists(String path) {
```
Lets keep BaseStorageClient FileSystem agnostic.

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageClient.java:
```diff
@@ -47,4 +47,14 @@ public interface StorageClient<T> {
+   * scheme. Scheme is not prefix for local and hdfs storage. See:
+   * https://github.com/linkedin/openhouse/issues/121
+   *
+   * @param path absolute path to a file including scheme
```
So this comment is not always true because the scheme may not have been passed along for hdfs files?
In that case, may be clarify that the path may or not have scheme specified. If scheme is not specified then it is assumed to be hdfs. For all other storage types, scheme must be specified.

vikrambohra replied:
Updated

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageManager.java:
```diff
@@ -93,4 +93,28 @@ public Storage getStorage(StorageType.Type storageType) {
+  public Storage getStorageFromPath(String path) {
+    for (Storage storage : storages) {
+      if (storage.isConfigured()) {
+        if (StorageType.LOCAL.equals(storage.getType())) {
```
Shouldn't this be fallback only if path does not start with any other configured endpoint?
What if the path is an S3 storage path but local storage is also configured for some other tables?

vikrambohra replied:
done. Please check

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/hdfs/HdfsStorageClient.java:
```diff
@@ -48,4 +49,20 @@ public FileSystem getNativeClient() {
   }
+
+  /**
+   * Checks if the path exists on the backend storage. Scheme is not prefix for local and hdfs
```
nit: I am having a hard time parsing the language "scheme is not prefix". Would you please consider paraphrasing to something like "Scheme is not specified in the path for local and hdfs storage".

jainlavina commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:
```diff
@@ -55,4 +58,26 @@ public S3Client getNativeClient() {
+  public boolean fileExists(String path) {
+    try {
+      HeadObjectRequest headObjectRequest =
+          HeadObjectRequest.builder().bucket(getRootPrefix()).key(path).build();
```
Did you check the headObject API spec that passing the full path including scheme and bucket name as key will work? It usually expects object name without the bucket name etc.

vikrambohra replied:
You are correct. Updated to parse bucket and key from path.

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageClient.java:
```diff
@@ -47,4 +47,14 @@ public interface StorageClient<T> {
+   * @param path absolute path to a file including scheme
+   * @return true if path exists else false
+   */
+  boolean fileExists(String path);
```
changes to exists(path)

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/hdfs/HdfsStorageClient.java:
```diff
@@ -48,4 +49,21 @@ public FileSystem getNativeClient() {
+    try {
+      return fs.exists(new Path(path));
+    } catch (IOException e) {
+      throw new RuntimeException("Exception checking path existence " + e.getMessage(), e);
```
done

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:
```diff
@@ -55,4 +58,26 @@ public S3Client getNativeClient() {
+      // Object does not exist
+      return false;
+    } catch (S3Exception e) {
+      throw new RuntimeException("Error checking S3 object existence: " + e.getMessage(), e);
```
done

vikrambohra commented on iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:
```diff
@@ -137,7 +138,11 @@ public void renameTable(TableIdentifier from, TableIdentifier to) {
+    return fileIOManager.getFileIO(resolveStorage(tableIdentifier).getType());
+  }
+
+  public Storage resolveStorage(TableIdentifier tableIdentifier) {
```
good catch. removed

vikrambohra commented on services/tables/src/test/java/com/linkedin/openhouse/tables/e2e/h2/SnapshotsControllerTest.java:
```diff
@@ -237,9 +239,15 @@ public void testPutSnapshotsReplicaTableType(GetTableResponseBody getTableRespon
+        "openhouse.tableLocation",
+        String.format(
+            "%s/%s/%s-%s/metadata.json",
+            "/tmp", getTableResponseBody.getDatabaseId(), getTableResponseBody.getTableId(), uuid));
```
done

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageClient.java:
```diff
@@ -47,4 +47,15 @@ public interface StorageClient<T> {
+   * @param path path to a file
+   * @return true if path exists else false
+   */
+  boolean fileExists(String path);
```
done

HotSushi commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -189,18 +188,26 @@ private Optional<UUID> extractUUIDFromRequestBody(
+      String manifestListPathString =
+          new Gson().fromJson(snapshotStr, JsonObject.class).get(manifestListKey).getAsString();
+      manifestListPathString =
+          StringUtils.removeStart(manifestListPathString, storage.getClient().getEndpoint());
```
is there a better way to achieve this without calling `storage.getClient().getEndpoint`, I'm ok with deferring this part to future pr.

HotSushi reviewed (commented):
For `extractUUIDFromTableProperties`, the logic before would check for directory existence thats missing now. We should be able to incorporate it as part of `Storage.isPathValid()`.

For `extractUUIDFromExistingManifestListPath ` we end up calling `storage.getClient().getEndpoint()`, we should introduce better layering/ or redefine the logic. But i'm ok pursuing this in a future PR.

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/Storage.java:
```diff
@@ -70,4 +70,15 @@ public interface Storage {
+   * @return true if endpoint is specified in cluster.yaml else false
+   */
+  default boolean isPathValid(String path) {
+    return path.startsWith(getClient().getEndpoint());
```
isPathValid should check three things:
- its appropriate for the right storage (ie: prefix matches s3://)
- its directory structure is valid
    - rootprefix is intact  (ie. /data/openhouse/db/tb/metadata.json has rootPrefix data/openhouse)
    - db/table directory is intact (ie. /data/openhouse/db/tb/metadata.json has directory structure thats created by storage.allocateTableLocation(), /db/tb-uuid)
- and object exists(ie. /data/openhouse/db/tb/metadata.json);

To achieve this, you might need to change the signature to: `boolean isPathValid(String path, dbname, tbname, tbuuid)`

vikrambohra commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -138,16 +135,20 @@ private void validatePathOfProvidedRequest(
-        extractFromTblPropsIfExists(databaseId + "." + tableId, tableProperties, DB_RAW_KEY);
-    String tblIdFromProps =
-        extractFromTblPropsIfExists(databaseId + "." + tableId, tableProperties, TBL_RAW_KEY);
+    String tableURI = String.format("%s.%s", databaseId, tableId);
```
@rohitkum2506 Gentle ping for your review.

HotSushi replied:
@Will-Lo / @maluchari is this something you can help with?

vikrambohra commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:
```diff
@@ -55,4 +62,39 @@ public S3Client getNativeClient() {
+    Preconditions.checkArgument(
+        path.startsWith(getEndpoint()), String.format("Invalid S3 URL format %s", path));
+    try {
+      URI uri = new URI(path);
```
I dont see a parseUri method on S3Utilities in v 2.20.18

vikrambohra commented:
Updated testing with latest changes

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/Storage.java:
```diff
@@ -70,4 +71,22 @@ public interface Storage {
+    String tableLocationPrefix = allocateTableLocation(databaseId, tableId, tableUUID, "");
+    Preconditions.checkArgument(
+        path.startsWith(tableLocationPrefix),
+        String.format("%s is not a prefix of table location %s", path, tableLocationPrefix));
```
maybe not the best idea to throw exception because it gets used in storage selection.

vikrambohra replied:
done

HotSushi commented on services/tables/src/main/java/com/linkedin/openhouse/tables/utils/TableUUIDGenerator.java:
```diff
@@ -138,16 +136,21 @@ private void validatePathOfProvidedRequest(
-    if (TableType.REPLICA_TABLE != tableType && !doesPathExist(previousPath)) {
-      log.error("Previous tableLocation: {} doesn't exist", previousPath);
+    if (TableType.REPLICA_TABLE != tableType
+        && !storage.isPathValid(dbIdFromProps, tblIdFromProps, tableUUIDProperty, tableLocation)) {
```
LGTM

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageManager.java:
```diff
@@ -93,4 +95,31 @@ public Storage getStorage(StorageType.Type storageType) {
+      String databaseId, String tableId, String tableUUID, String path) {
+    for (Storage storage : storages) {
+      if (storage.isConfigured()) {
+        if (storage.isPathValid(databaseId, tableId, tableUUID, path)) {
```
maybe:
```
public Storage getStorageFromPath(path) {
 ...  
 path.beginsWith(getCLient().getEndpoint()
....
}
```

vikrambohra replied:
done

HotSushi commented on cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/StorageManager.java:
```diff
@@ -93,4 +95,32 @@ public Storage getStorage(StorageType.Type storageType) {
+        if (path.startsWith(storage.getClient().getEndpoint())) {
+          log.info("Resolved to {} storage for path {}", storage.getType().toString(), path);
+          return storage;
+        } else if (path.startsWith("/")) {
```
can we do `storage.storageType = StorageType.HDFS` and path.startsWith("/")? 

Current condition is weak and might accidentally greenlight incorrect storage

vikrambohra replied:
done

HotSushi reviewed (commented):
Looks good after one nit.

HotSushi reviewed (approved):
LGTM. Lets do a docker test before merging

vikrambohra commented:
> LGTM. Lets do a docker test before merging

Test works. Added in testing done section
